File: scala_study/src/main/scala/leetcode/detectCycle.scala
package leetcode

class ListNode(_x: Int, _next: ListNode) {
  var x = _x
  var next = _next

}

object Solution{
  def detectCycle(head: ListNode):ListNode = {
    var slow = head
    var fast = slow
    while (fast != null){
      slow=slow.next
      if (fast.next == null) return null
      fast = fast.next.next

      if (fast == slow){
        var cur = head
        while (cur != slow){
          cur =cur.next
          slow = slow.next
        }
        return cur
      }
    }
    null
  }
}

File: JavaLearn/src/com/leetcode/DetectCycle.java
package com.leetcode;

public class DetectCycle {
    public ListNode detectCycle(ListNode head) {

        if (head == null){
            return head;
        }

        ListNode ans = null;

        ListNode slow = head, fast = head;

        while (fast != null){
            slow = slow.next;
            if (fast.next != null){
                fast = fast.next.next;
            }else {
                return null;
            }

            if (slow == fast){
                ans = head;

                while (ans != slow){
                    ans=ans.next;
                    slow = slow.next;
                }
                return ans;
            }
        }


        return ans;
    }
}
//class ListNode {
//    int val;
//    ListNode next;
//    ListNode(int x) {
//        val = x;
//        next = null;
//    }
//}